packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -0,0 +1,334 @@
+	}
+
+	// Slice the trailing `, ` from the `result`.
+	result = result.slice(0, -2);
+
+	return `{ ${result} }`;
+}
+
+/**
+ * Returns the allowed fields & types for the leaf fields.
+ */
+function getLeafAllowedTypes(schema: TreeFieldStoredSchema): string {
+	let result = "";
+	const leafTypes = schema.types;
+
+	if (leafTypes === undefined) {
+		result = "any";
+	} else {
+		for (const type of leafTypes) {
+			result += `${type} | `;
+		}
+	}
+
+	// Slice the trailing ` | ` from the `result`.
+	result = `${result.slice(0, -3)}`;
+
+	return result;
+}
+
+/**
+ * Returns the schema & leaf value of the node with type {@link LeafNodeStoredSchema}.
+ */
+function visualizeLeafNodeStoredSchema(
+	tree: JsonableTree,
+	allowedTypes: string,
+): SharedTreeLeafNode {
+	return {
+		schema: {
+			schemaType: SharedTreeSchemaType.LeafNodeStoredSchema,
+			allowedTypes,
+		},
+		value: JSON.stringify(tree.value),
+		kind: VisualSharedTreeNodeKind.LeafNode,
+	};
+}
+
+/**
+ * Returns the schema & fields of the node with type {@link ObjectNodeStoredSchema}.
+ */
+function visualizeObjectNodeStoredSchema(
+	tree: JsonableTree,
+	schema: ObjectNodeStoredSchema,
+	contentSnapshot: SharedTreeContentSnapshot,
+): VisualSharedTreeNode {
+	const treeFields = tree.fields;
+
+	if (treeFields === undefined || Object.keys(treeFields).length === 0) {
+		return {
+			schema: {
+				name: tree.type,
+				schemaType: SharedTreeSchemaType.ObjectNodeStoredSchema,
+				allowedTypes: getObjectAllowedTypes(schema),
+			},
+			fields: {},
+			kind: VisualSharedTreeNodeKind.InternalNode,
+		};
+	}
+
+	const fields: Record<string | number, VisualSharedTreeNode> = {};
+
+	// If the child node is a leaf node, get the allowed types from the parent schema before entering {@link visualizeSharedTreeNodeBySchema}.
+	let leafAllowedTypes;
+
+	// {@link EmptyKey} indicates an array field (e.g., `schemabuilder.array()`).
+	// Hides level of indirection by omitting the empty key in the visual output.
+	if (
+		Object.keys(treeFields).length === 1 &&
+		Object.prototype.hasOwnProperty.call(treeFields, EmptyKey)
+	) {
+		const children = treeFields[EmptyKey];
+
+		for (let i = 0; i < children.length; i++) {
+			const childSchema = contentSnapshot.schema.nodeSchema.get(children[i].type);
+
+			// If the node within the array is a leaf node, get the allowed types from the parent schema.
+			if (childSchema instanceof LeafNodeStoredSchema) {
+				const parentSchema = schema.objectNodeFields;
+
+				for (const [, leafSchema] of parentSchema) {
+					leafAllowedTypes = getLeafAllowedTypes(leafSchema);
+				}
+			}
+
+			fields[i] = visualizeSharedTreeNodeBySchema(
+				children[i],
+				childSchema,
+				contentSnapshot,
+				leafAllowedTypes as string,
+			);
+		}
+	} else {
+		for (const [fieldKey, childField] of Object.entries(treeFields)) {
+			assert(
+				childField.length === 1,
+				"Non-array node should not have more than one child field.",
+			);
```

Comment:
Not sure this kind of validation (all the asserts in this file) belongs in Devtools. Tree itself should be doing it when appropriate, and I think here we should just trust that the data is correct. This is also the only reason we had to add the core-utils dependency, so I'd remove it together with this. @Josmithr any thoughts?

Reply:
We could replace the asserts with console errors / warnings in places where the violations aren't critical. In places where we would otherwise not be able to craft a valid `VisualTree` from the input, we should create an `UnknownContent` tree so that developers can see something in the UI.

Reply:
I have removed the `core-utils` package and replaced it with the `throw new Error`